File: rust/load_balancer/src/least_response_load_balancer.rs
use crate::backend::Backend;
use crate::health::Health;
use crate::internal_error::InternalError;
use crate::load_balancer::LoadBalancer;
use crate::min_heap_item::MinHeapItem;

use async_trait::async_trait;
use log::{error, info, warn};
use std::collections::BinaryHeap;
use tokio::sync::RwLock as TokioRwLock;

/// Represents a very basic load balancer. Sends the requests to healthy backend servers in a round
/// robin fashion.
#[derive(Debug)]
pub struct LeastResponseLoadBalancer {
    /// List of unhealthy backends servers
    unhealthy_backends: TokioRwLock<Vec<Box<dyn Backend>>>,

    /// Min heap of healthy backend servers. The heap is ordered by the response time of the
    /// backends
    healthy_backends: TokioRwLock<BinaryHeap<MinHeapItem<Box<dyn Backend>>>>,
}

impl LeastResponseLoadBalancer {
    /// Creates a new load balancer with the given list of backend servers to route the requests
    /// to.
    pub fn new(backends: Vec<Box<dyn Backend>>) -> Self {
        let mut healthy_backends = BinaryHeap::new();
        for backend in backends.into_iter() {
            healthy_backends.push(MinHeapItem {
                priority: 0.0,
                element: backend,
            });
        }
        Self {
            unhealthy_backends: TokioRwLock::new(Vec::new()),
            healthy_backends: TokioRwLock::new(healthy_backends),
        }
    }
}

#[async_trait]
impl LoadBalancer for LeastResponseLoadBalancer {
    // Returns the next available backend server to which the request can be sent. If none are
    // available, an error is returned.
    async fn next_available_backend(&self) -> Result<Box<dyn Backend>, String> {
        let r_healthy_backends = self.healthy_backends.read().await;
        if r_healthy_backends.is_empty() {
            return Err("No backend server available".to_string());
        }

        let MinHeapItem { element, .. } = r_healthy_backends.peek().unwrap();

        Ok(element.clone())
    }

    async fn send_request(&self) -> Result<String, InternalError> {
        let mut w_healthy_backends = self.healthy_backends.write().await;
        if w_healthy_backends.is_empty() {
            return Err(InternalError::NoBackendAvailable);
        }

        let MinHeapItem {
            element: backend, ..
        } = w_healthy_backends.pop().unwrap();

        // Send the request to the backend server
        let response = backend.send_request().await;
        match response {
            Ok(r) => {
                info!("{:?}", r);
                w_healthy_backends.push(MinHeapItem {
                    priority: backend.response_time_ms().await,
                    element: backend,
                });
                drop(w_healthy_backends);
                let body = r.text_with_charset("utf-8").await.unwrap();
                Ok(body)
            }
            Err(e) => {
                error!(
                    "Failed to send request to backend server: {:?}, trying next one",
                    e
                );
                let mut w_unhealthy_backends = self.unhealthy_backends.write().await;
                w_unhealthy_backends.push(backend);
                drop(w_unhealthy_backends);
                drop(w_healthy_backends);

                self.send_request().await
            }
        }
    }

    /// Checks and update the health status of all backend servers.
    async fn check_backends_healths(&self) {
        // This is used for profiling only
        let start_time = std::time::Instant::now();

        let mut new_healthy_backends = BinaryHeap::new();
        let mut new_unhealthy_backends: Vec<Box<dyn Backend>> = Vec::new();

        let mut w_healthy_backends = self.healthy_backends.write().await;
        // check healthy backends
        while let Some(MinHeapItem {
            element: backend, ..
        }) = w_healthy_backends.pop()
        {
            backend.check_health().await;
            if backend.health().await == Health::Healthy {
                let response_time = backend.response_time_ms().await;
                info!(
                    "Backend {:?} is healthy with response time {}ms",
                    backend, response_time
                );
                new_healthy_backends.push(MinHeapItem {
                    priority: response_time,
                    element: backend,
                });
            } else {
                warn!("Backend {:?} is unhealthy", backend);
                new_unhealthy_backends.push(backend);
            }
        }

        // check unhealthy backends
        let mut w_unhealthy_backends = self.unhealthy_backends.write().await;
        while let Some(backend) = w_unhealthy_backends.pop() {
            backend.check_health().await;
            if backend.health().await == Health::Healthy {
                info!("Backend {:?} is now healthy", backend);
                new_healthy_backends.push(MinHeapItem {
                    priority: backend.response_time_ms().await,
                    element: backend,
                });
            } else {
                info!("Backend {:?} is still unhealthy", backend);
                new_unhealthy_backends.push(backend);
            }
        }

        *w_healthy_backends = new_healthy_backends;
        *w_unhealthy_backends = new_unhealthy_backends;
        let healthy_backends_count = w_healthy_backends.len();
        let unhealthy_backends_count = w_unhealthy_backends.len();

        let best_backend = w_healthy_backends.peek().unwrap();
        let best_backend_priority = best_backend.priority;
        let best_backend_address: String = best_backend.element.address().into();

        drop(w_healthy_backends);
        drop(w_unhealthy_backends);

        // For profiling only, measures how much time it took to check all backends health
        let end_time = std::time::Instant::now();
        let elapsed_time = end_time.duration_since(start_time).as_millis();
        info!("checking all backends health took {}ms", elapsed_time);
        info!(
            "Best backend: {}ms, {}",
            best_backend_priority, best_backend_address
        );
        info!(
            "Healthy backends: {}, Unhealthy backends: {}",
            healthy_backends_count, unhealthy_backends_count
        );
    }
}

File: cpp/source/least_response_load_balancer.cpp
#include <least_response_load_balancer.h>

namespace load_balancer {

LeastResponseLoadBalancer::LeastResponseLoadBalancer(std::vector<std::shared_ptr<Backend>> backends,
                                                     int health_check_interval_s)
    : health_check_interval_s{health_check_interval_s},
      current_backend_index{0},
      health_check_thread_running{false},
      backend_semaphore{1} {
  spdlog::info("Load balancer started with {} backends", backends.size());
  for (auto &backend : backends) {
    spdlog::info("Backend at {} with response time {}", backend->address(),
                 backend->response_time().count());
    this->healthy_backends.push(backend);
  }
}

drogon::Task<void> LeastResponseLoadBalancer::check_backend_healths() {
  spdlog::info("Checking health of all backends");

  auto start = std::chrono::high_resolution_clock::now();

  std::priority_queue<std::shared_ptr<Backend>, std::vector<std::shared_ptr<Backend>>,
                      MinHeapBackendComparator>
      new_healthy_backends;
  std::vector<std::shared_ptr<Backend>> new_unhealthy_backends;

  this->backend_semaphore.acquire();

  spdlog::info("Checking health of {} healthy backends", this->healthy_backends.size());
  while (!this->healthy_backends.empty()) {
    auto backend = this->healthy_backends.top();
    this->healthy_backends.pop();

    co_await backend->check_health();
    if (backend->health() == Health::Healthy) {
      new_healthy_backends.push(backend);
    } else {
      spdlog::warn("Backend at {} is unhealthy", backend->address());
      new_unhealthy_backends.push_back(backend);
    }
  }

  spdlog::info("Checking health of {} unhealthy backends", this->unhealthy_backends.size());
  for (auto backend : this->unhealthy_backends) {
    co_await backend->check_health();
    if (backend->health() == Health::Healthy) {
      spdlog::info("Backend at {} is now healthy", backend->address());
      new_healthy_backends.push(backend);
    } else {
      spdlog::warn("Backend at {} is still unhealthy", backend->address());
      new_unhealthy_backends.push_back(backend);
    }
  }

  this->healthy_backends = new_healthy_backends;
  this->unhealthy_backends = new_unhealthy_backends;

  this->backend_semaphore.release();

  auto end = std::chrono::high_resolution_clock::now();
  auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
  spdlog::info("Health check of all backends took {}ms", duration.count());

  co_return;
}

void LeastResponseLoadBalancer::start_health_checks() {
  this->health_check_thread_running = true;
  spdlog::info("Starting health checks every {}s", this->health_check_interval_s);

  drogon::async_run([this]() -> drogon::Task<void> {
    while (this->health_check_thread_running) {
      co_await this->check_backend_healths();
      co_await drogon::sleepCoro(trantor::EventLoop::getEventLoopOfCurrentThread(),
                                 std::chrono::seconds(this->health_check_interval_s));
    }
    spdlog::info("Stopped health checks");
    co_return;
  });
}

void LeastResponseLoadBalancer::stop_health_checks() {
  spdlog::info("Stopping health checks");
  this->health_check_thread_running = false;
}

drogon::Task<drogon::HttpResponsePtr> LeastResponseLoadBalancer::send_request(
    drogon::HttpRequestPtr request) {
  if (this->healthy_backends.empty()) {
    spdlog::error("No healthy backends available");
    throw std::runtime_error("No healthy backends available");
  }

  this->backend_semaphore.acquire();

  auto backend = this->healthy_backends.top();
  this->healthy_backends.pop();

  auto response = co_await backend->send_request(request);
  if (response->getStatusCode() < drogon::HttpStatusCode::k200OK
      || response->getStatusCode() > drogon::HttpStatusCode::k206PartialContent) {
    spdlog::error("Backend at {} returned status code {}", backend->address(),
                  static_cast<int>(response->getStatusCode()));
    this->unhealthy_backends.push_back(backend);
  } else {
    this->healthy_backends.push(backend);
  }

  this->backend_semaphore.release();

  co_return response;
}

} /* namespace load_balancer */
